contracts/governed_token_wrapper/lib.rs

```diff
@@ -0,0 +1,219 @@
+                instance.proposal_nonce = proposal_nonce;
+            })
+        }
+
+        /// Used to wrap tokens on behalf of a sender.
+        ///
+        /// token_address is the address of PSP22 to transfer to, if token_address is None,
+        /// then it's a Native token address
+        ///
+        /// amount is the amount of token to transfer
+        #[ink(message, payable)]
+        pub fn wrap(&mut self, token_address: Option<AccountId>, amount: Balance) {
+            self.is_valid_wrapping(token_address, amount);
+
+            // determine amount to use
+            let amount_to_use = if token_address.is_none() {
+                self.env().transferred_value()
+            } else {
+                amount
+            };
+
+            let cost_to_wrap = self.get_fee_from_amount(amount_to_use);
+
+            let leftover = amount_to_use.saturating_mul(cost_to_wrap);
+
+            if token_address.is_none() {
+                // mint the native value sent to the contract
+                self.mint(self.env().caller(), leftover);
+
+                // transfer costToWrap to the feeRecipient
+                if self
+                    .env()
+                    .transfer(self.fee_recipient, cost_to_wrap)
+                    .is_err()
+                {
+                    panic!("{}", ERROR_MSG);
+                }
+            } else {
+                // psp22 transfer of  liquidity to token wrapper contract
+                self.transfer_from(
+                    self.env().caller(),
+                    self.env().account_id(),
+                    leftover,
+                    Vec::<u8>::new(),
+                )
+                .is_ok();
+
+                // psp22 transfer to fee recipient
+                self.transfer_from(
+                    self.env().caller(),
+                    self.fee_recipient,
+                    cost_to_wrap,
+                    Vec::<u8>::new(),
+                )
+                .is_ok();
+
+                // mint the wrapped token for the sender
+                self.mint(self.env().caller(), leftover);
+            }
+        }
+
+        /// Checks to determine if it's safe to wrap
+        fn is_valid_wrapping(
+            &mut self,
+            token_address: Option<AccountId>,
+            amount: Balance,
+        ) -> Result<()> {
+            if token_address.is_none() {
+                if amount == 0 {
+                    return Err(Error::InvalidAmountForNativeWrapping);
+                }
+
+                if !self.is_native_allowed {
+                    return Err(Error::NativeWrappingNotAllowed);
+                }
+            } else {
+                if self.env().transferred_value() == 0 {
+                    return Err(Error::InvalidValueSentForWrapping);
+                }
+
+                if !self.is_valid_address(token_address.unwrap()) {
+                    return Err(Error::InvalidTokenAddress);
+                }
+            }
```

Comment:
Seems like the logic of this part is a little bit wrong.
`if token_address.is_none()` then it should check `self.env().transferred_value()` for native currency.

Reply:
I think it is actually right, it's same as what goes on in the [solidity](https://github.com/webb-tools/protocol-solidity/blob/main/contracts/tokens/TokenWrapper.sol#L241 ) end 
Moreso, i am using the native self.env().transferred_value() [here]( https://github.com/webb-tools/protocol-ink/blob/dami/governed-token-wrapper/contracts/governed_token_wrapper/lib.rs#L151)
What is happening here is just a validity check

Reply:
Please check again.
If `token_address.is_none()`, then it is for checking the `native currency`, which should check with `self.env().transferred_value()`.
I think line 206 - 208 and line 214 - 216 should be swapped.

Reply:
I want to ask a question, is msg.value() in solidity end for checking native currency? @nepoche

Reply:
yes, `msg.value` is for checking `ether(native currency)` in solidity.

Reply:
If so, then this  [check](https://github.com/webb-tools/protocol-solidity/blob/main/contracts/tokens/TokenWrapper.sol#L241) in solidity is wrong then, right @duguorong009 @nepoche

Reply:
Alright. I totally get you now. Thanks